torchtune/modules/model_fusion/_fusion.py

```diff
@@ -477,3 +477,107 @@ def forward(
             input_pos=input_pos,
         )
         return output
+
+class EarlyFusionModel(nn.Module):
+    def __init__(
+        self,
+        decoder: TransformerDecoder,
+        encoders: nn.ModuleDict,
+        encoder_tokens: Dict[str, int],
+        decoder_trainable: bool,
+        encoders_trainable: Dict[str, bool],
+    ):
+        super().__init__()
+        self.decoder = decoder
+        self.encoders = encoders
+        self.encoder_tokens = encoder_tokens
+
+        # A little surgery in the decoder to give the
+        # fusion module access to control the embeddings
+        # The alternative is to pass a special tok_embeddings
+        # module into TransformerDecoder builder that does the
+        # merging there
+        self.tok_embeddings = decoder.tok_embeddings
+        decoder.tok_embeddings = nn.Identity()
+
+        self.register_state_dict_post_hook(self._state_dict_hook)
+        self.register_load_state_dict_pre_hook(
+            self._load_state_dict_hook,
+            with_module=True
+        )
+
+        trainable_params = set()
+        for encoder, trainable in encoders_trainable.items():
+            if trainable:
+                trainable_params |= {
+                    f"encoders.{encoder}.{n}" for n, p in self.encoders[encoder].named_parameters()
+                }
+        if decoder_trainable:
+            trainable_params |= {
+                f"decoder.{n}" for n, p in self.decoder.named_parameters()
+            }
```

Comment:
This is missing the logic and the parameter to make fusion modules trainable/untrainable